### PR TITLE
Add a function to detect Asset Types

### PR DESCRIPTION
**Pull request by seilagamo**
Add a function that given an identifier automatically detects the Asset Types.

### CODE REVIEW

File: types.go [hunk 1]
```diff
@@ -199,3 +200,71 @@ func IsHostname(target string) bool {
 
 	return len(r) > 0
 }
+
+// DetectAssetTypes detects the asset types from a identifier.
+func DetectAssetTypes(identifier string) ([]string, error) {
+	if IsAWSARN(identifier) {
```
Comment (jroimartin) [inline types.go:206]: It is a bit weird that any valid AWS ARN is considered an AWS account. Should we create a function `IsAWSAccount` that adds extra validations?

File: types.go [hunk 2]
```diff
@@ -199,3 +200,71 @@ func IsHostname(target string) bool {
...
+		isWeb = true
+
+		// From a URL like https://adevinta.com not only a WebAddress
+		// type can be extracted, also a hostname (adevinta.com) and
+		// potentially a domain name.
+		u, err := url.ParseRequestURI(identifier)
+		if err != nil {
+			return nil, err
+		}
+		identifier = u.Hostname() // Overwrite identifier to check for hostname and domain.
```
Comment (jroimartin) [inline types.go:247]: Move comment to previous line.

**Review by jroimartin — CHANGES_REQUESTED** [review]

File: types_test.go [hunk 3]
```diff
@@ -588,23 +588,79 @@ func TestIsHostname(t *testing.T) {
...
+			identifier:     "finntech/docker-elasticsearch-kubernetes",
+			wantAssetTypes: nil,
+			wantNilErr:     true},
+		{
+			name:           "valid hostname and web address",
+			identifier:     "https://www.example.com",
+			wantAssetTypes: []AssetType{Hostname, WebAddress},
+			wantNilErr:     true},
+		{
+			name:           "valid artifact",
```
Comment (jroimartin) [inline types_test.go:660]: What about `valid docker image v2 spec`? Based on:

https://github.com/adevinta/vulcan-types/blob/189dc9100670916216ce9e120f64bebe9644b485/types.go#L86

File: types.go [hunk 4]
```diff
@@ -201,37 +201,83 @@ func IsHostname(target string) bool {
 	return len(r) > 0
 }
 
-// DetectAssetTypes detects the asset types from a identifier.
-func DetectAssetTypes(identifier string) ([]string, error) {
+type AssetType string
+
+// Asset Types for vulcan assets.
```
Comment (jroimartin) [inline types.go:206]: s/Types/types/

**Review by jroimartin — CHANGES_REQUESTED** [review]
Overall LGTM. I just proposed a few minor changes.

File: types.go (hunk 1 above)
Comment (jroimartin) [reply types.go:252]: @manelmontilla @julianvilas @jesusfcr what do you think?

The current logic is exactly the same used by vulcan-api in [github.com/adevinta/vulcan-api/pkg/api/service.getTypesFromIdentifier](https://github.com/adevinta/vulcan-api/blob/496a7a472a648c975d14640c6bdd160ef0ea1767/pkg/api/service/assets.go#L680).

**Review by jroimartin — APPROVED** [review]
LGTM. However, let's wait to solve the [comment](https://github.com/adevinta/vulcan-types/pull/85#discussion_r1263475130) about AWS accounts detection before merging.

File: types.go (hunk 1 above)
Comment (jesusfcr) [reply types.go:252]: I think we should restrict the arn to just accounts, as this accepts any arn.
arn:aws:sqs:eu-west-1:892351162211:HermesAlertsSQS

Comment (jesusfcr) [reply types.go:252]: Regarding to the overall pr I think it's great to have this unified in one place, I already missed this when working in `vulcan-local` so I decided to clone and adapt the code.

Comment (manelmontilla) [reply types.go:252]: I agree that is weird, I guess there was a reason for doing that way but I honestly I don't remember it. I also agree with @jesusfcr in the sense that we must restrict the account ARN's and nothing else (I'm giving by granted that we want to use the format of the root user of an account: `arn:aws:iam::XXXX:root` as the only valid format for an AWS account) . In any case, I would not include this modification in this PR.

File: types.go [hunk 5]
```diff
@@ -79,15 +79,16 @@ func IsAWSARN(target string) bool {
...
-//   Valid: registry.hub.docker.com/path1/path2/artifact (compliant with V2 spec)
-//   Valid: registry.hub.docker.com/artifact (compliant with V2 spec)
-//   Valid: localhost:5500/library/debian
-//   Not valid: metasploitframework/metasploit-framework:latest
-//   Not valid: metasploitframework/metasploit-framework
-//   Not valid: debian
+//
+//   - Valid: registry.hub.docker.com/metasploitframework/metasploit-framework:latest
+//   - Valid: registry.hub.docker.com/metasploitframework/metasploit-framework
+//   - Valid: registry.hub.docker.com/library/debian
```
Comment (jesusfcr) [inline types.go:85]: I would add an example with `docker.io` and perhaps some with other big registries like quay/ghcr.io, just to make it clear.

File: types.go (hunk 1 above)
Comment (jroimartin) [reply types.go]: Please check 5f0316d.

Comment (jroimartin) [reply types.go]: @manelmontilla sorry, I read your comment just after pushing the fix. Given that we are not introducing breaking changes, I think it makes sense keep the commit and have the change available for other packages.

I assumed the format `arn:aws:iam::XXXX:root` based on [AWS documentation](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html#principal-accounts).

Of course, I'm fine with dropping the commit and add it in a different PR.

Comment (manelmontilla) [reply types.go]: No, okey as it is already done and as you said, it doesn't introduce breaking changes, I think it's worth keeping.

File: types.go (hunk 5 above)
Comment (seilagamo) [reply types.go:96]: I have added the examples and tests in commit 7222b2ab44ff1e7d6ea4f55d615c3a8c374a7353

**Review by jesusfcr — APPROVED** [review]

**Pull request merged**